barretenberg/cpp/src/barretenberg/vm2/tracegen/sha256_trace.cpp

```diff
@@ -0,0 +1,355 @@
+
+    // Compute ch := (e & f) ^ (~e & g);
+    // Compute ~e
+    uint32_t not_e = ~state[4];
+    trace.set(C::sha256_not_e, row, not_e);
+    // Compute e & f
+    uint32_t e_and_f = state[4] & state[5];
+    trace.set(C::sha256_e_and_f, row, e_and_f);
+    // Compute ~e & g
+    uint32_t not_e_and_g = not_e & state[6];
+    trace.set(C::sha256_not_e_and_g, row, not_e_and_g);
+    // Compute (e & f) ^ (~e & g)
+    uint64_t ch = e_and_f ^ not_e_and_g;
+    trace.set(C::sha256_ch, row, ch);
+
+    // Compute S0 := ror(a, 2U) ^ ror(a, 13U) ^ ror(a, 22U);
+    // Compute ror(a, 2)
+    uint32_t rot_2 = ror_with_witness(state[0], 2, trace, C::sha256_a_rotr_2, C::sha256_lhs_a_2, C::sha256_rhs_a_2);
+    // Compute ror(a, 13)
+    uint32_t rot_13 =
+        ror_with_witness(state[0], 13, trace, C::sha256_a_rotr_13, C::sha256_lhs_a_13, C::sha256_rhs_a_13);
+    // Compute ror(a, 22)
+    uint32_t rot_22 =
+        ror_with_witness(state[0], 22, trace, C::sha256_a_rotr_22, C::sha256_lhs_a_22, C::sha256_rhs_a_22);
+
+    // Compute ror(a, 2) ^ ror(a, 13)
+    trace.set(C::sha256_a_rotr_2_xor_a_rotr_13, row, rot_2 ^ rot_13);
+    // Compute S0, this can't overflow but we expand to uint64_t for later use
+    uint64_t S0 = rot_2 ^ rot_13 ^ rot_22;
+    trace.set(C::sha256_s_0, row, S0);
+
+    // Compute Maj := (a & b) ^ (a & c) ^ (b & c);
+    // Compute a & b
+    uint32_t a_and_b = state[0] & state[1];
+    trace.set(C::sha256_a_and_b, row, a_and_b);
+    // Compute a & c
+    uint32_t a_and_c = state[0] & state[2];
+    trace.set(C::sha256_a_and_c, row, a_and_c);
+    // Compute b & c
+    uint32_t b_and_c = state[1] & state[2];
+    trace.set(C::sha256_b_and_c, row, b_and_c);
+    // Compute (a & b) ^ (a & c)
+    trace.set(C::sha256_a_and_b_xor_a_and_c, row, a_and_b ^ a_and_c);
+    // Compute Maj, this is expanded to uint64_t to detect later overflows
+    uint64_t maj = a_and_b ^ a_and_c ^ b_and_c;
+    trace.set(C::sha256_maj, row, maj);
+
+    // Compute temp values, these need be 64-bit integers and performed modulo 2^32
+    uint64_t temp1 = static_cast<uint64_t>(state[7]) + S1 + ch + round_constant + round_w;
+    uint64_t temp2 = S0 + maj;
+    uint64_t next_a = temp1 + temp2;
+    into_limbs_with_witness(next_a, 32, trace, C::sha256_next_a_lhs, C::sha256_next_a_rhs);
+    trace.set(C::sha256_round_constant, row, round_constant);
+    uint32_t a = static_cast<uint32_t>(next_a);
+
+    // Additions can overflow 32 bits so we perform modulo reduction
+    uint64_t next_e = state[3] + temp1;
+    into_limbs_with_witness(next_e, 32, trace, C::sha256_next_e_lhs, C::sha256_next_e_rhs);
+    uint32_t e = static_cast<uint32_t>(next_e);
+
+    return {
+        a,        /*a = temp1 + temp2*/
+        state[0], /*b = a*/
+        state[1], /*c = b*/
+        state[2], /*d = c*/
+        e,        /*e = d + temp1*/
+        state[4], /*f = e*/
+        state[5], /*g = f*/
+        state[6], /*h = g*/
+    };
+}
+
```

Comment:
I will trust you for most of this. But indeed you effort paid off. It's looking much better now!